For this assignment, we will explore Twitter Emotion Classification. The goal is the identify the primary emotion expressed in a tweet. Consider the following tweets:
```
Tweet 1: @NationalGallery @ThePoldarkian I have always loved this painting.
Tweet 2: '@tateliverpool #BobandRoberta: I am angry more artists that have a profile are not speaking up #foundationcourses.'
``` 

How would you describe the emotions in `Tweet 1` vs `Tweet 2`? `Tweet 1` expresses enjoyment and happiness, while `Tweet 2` directly expresses anger. For this assignment, we will be working with the SMILE Twitter Emotion Dataset ([Wang et al. 2016](https://ceur-ws.org/Vol-1619/paper3.pdf)). At a high level, our goal is to develop different models (rule-based, machine learning, and deep learning), which can be used to identify the emotion of a tweet. You will be required to clean and preprocess the data, generate features for classification, train various models, and evaluate the models. 


*Submission Details*
Please complete all the tasks in “Assignment 1.ipynb” and upload your submission as a Python notebook on Blackboard with the filename “StudentID_Lastname.ipynb”. Assignment 1 will be due by 11:59 PM GMT Monday February 20th, 2023.  

*Grading Policy*

Assignment 1 is graded and will be worth 25% of your overall grade. This assignment is worth a total of 50 points distributed over the tasks below. 
Please note that this is an individual assignment and you must not work with other students to complete this assessment. Any copying from other students, from student exercises from previous years, and any internet resources will not be tolerated. Plagiarised assignments will receive zero marks and the students who commit this act will be reported. 

Feel free to reach out to the TAs and instructors if you have any questions.

Before you get started, run the cell below to download the dataset into memory and a few relevant libraries.

In [1]:
!wget -O data.csv "https://figshare.com/ndownloader/files/4988956"
!pip install emoji

import nltk
nltk.download('punkt')

--2023-02-18 16:45:43--  https://figshare.com/ndownloader/files/4988956
Resolving figshare.com (figshare.com)... 54.155.235.28, 54.155.203.252, 2a05:d018:1f4:d003:e231:deac:908b:46d0, ...
Connecting to figshare.com (figshare.com)|54.155.235.28|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/4988956/smileannotationsfinal.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230218/eu-west-1/s3/aws4_request&X-Amz-Date=20230218T164543Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=733796fb46e3735030bc6773ba69f0646d1b25dc3ca9a2d3878618384000b118 [following]
--2023-02-18 16:45:43--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/4988956/smileannotationsfinal.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230218/eu-west-1/s3/aws4_request&X-Amz-Date=20230218T164543Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=733796fb46e373503

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [2]:
import pandas as pd
data = pd.read_csv("data.csv", names=["id", "tweet", "label"])
display(data.head(5))

id                                              tweet  \
0  611857364396965889  @aandraous @britishmuseum @AndrewsAntonio Merc...   
1  614484565059596288  Dorian Gray with Rainbow Scarf #LoveWins (from...   
2  614746522043973632  @SelectShowcase @Tate_StIves ... Replace with ...   
3  614877582664835073  @Sofabsports thank you for following me back. ...   
4  611932373039644672  @britishmuseum @TudorHistory What a beautiful ...   

    label  
0  nocode  
1   happy  
2   happy  
3   happy  
4   happy

In [3]:
import numpy as np
import nltk
!pip install transformers >> NULL
!pip install torch

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


## Task 1. Data Cleaning, Preprocessing, and splitting [15 points]
The `data` environment contains the SMILE dataset loaded into a pandas dataframe object. Our dataset has three columns: id, tweet, and label. The `tweet` column contains the raw scraped tweet and the `label` column contains the annotated emotion category. Each tweet is labelled with one of the following emotion labels:
- 'nocode', 'not-relevant' 
- 'happy', 'happy|surprise', 'happy|sad'
- 'angry', 'disgust|angry', 'disgust' 
- 'sad', 'sad|disgust', 'sad|disgust|angry' 
- 'surprise'

### Task 1a. Label Consolidation [ 3 points]
As we can see above the annotated categories are complex. Several tweets express complex emotions like (e.g. 'happy|sad') or multiple emotions (e.g. 'sad|disgust|angry'). The first things we need to do is clean up our dataset by removing complex examples and consolidating others so that we have a clean set of emotions to predict. 

For Task 1a., write code which does the following:
1. Drops all rows which have the label "happy|sad", "happy|surprise", 'sad|disgust|angry', and 'sad|angry'.
2. Re-label 'nocode' and 'not-relevant' as 'no-emotion'.
3. Re-label 'disgust|angry' and 'disgust' as 'angry'.
4. Re-label 'sad|disgust' as 'sad'.

Your updated `data' dataframe should have 3,062 rows and 5 label categories (no-emotion, happy, angry, sad, and surprise).


In [4]:
data = data[data["label"].isin(["happy|sad", "happy|surprise", 'sad|disgust|angry', 'sad|angry']) == False]
data['label'].mask(data['label'] == 'nocode', "no-emotion", inplace=True)
data['label'].mask(data['label'] == 'not-relevant', "no-emotion", inplace=True)
data['label'].mask(data['label'] == 'disgust|angry', "angry", inplace=True)
data['label'].mask(data['label'] == 'disgust', "angry", inplace=True)
data['label'].mask(data['label'] == 'sad|disgust', "sad", inplace=True)

/usr/local/lib/python3.8/dist-packages/pandas/core/generic.py:8870: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


### Task 1a Tests 
Run the cell below to evaluate your code. To get full credit for this task, your code must pass all tests. Any alteration of the testing code will automatically result in 0 points. 

In [5]:
# Test 1. Data should have 5 unique labels.
print(f"Unique label test: {len(data['label'].unique()) == 5}")

# Test 2. Data labels must be: angry, happy, no-emotion, sad, and surprise
labels = ["angry", "happy", "no-emotion", "sad", "surprise"]
print(f"Label check: { set(data['label'].unique()).difference(labels) == set() }")

# Test 3. Check example counts per label
print(f"Angry example count: {len(data[data['label']=='angry']) == 70}")
print(f"Happy example count: {len(data[data['label']=='happy']) == 1137}")
print(f"No-Emotion example count: {len(data[data['label']=='no-emotion']) == 1786}")
print(f"Sad example count: {len(data[data['label']=='sad']) == 34}")
print(f"Surprise example count: {len(data[data['label']=='surprise']) == 35}")

Unique label test: True
Label check: True
Angry example count: True
Happy example count: True
No-Emotion example count: True
Sad example count: True
Surprise example count: True


### Task 1b. Tweet Cleaning and Processing [10 points]
Raw tweets are noisy. Consider the example below: 
```
'@tateliverpool #BobandRoberta: I am angry more artists that have a profile are not speaking up #foundationcourses. 😠'
```
The mention @tateliverpool and hashtag #BobandRoberta are extra noise that don't directly help with understanding the emotion of the text. The accompanying emoji can be useful but needs to be decoded to it text form :angry: first. 

For this task you will fill complete the `preprocess_tweet` function below with the following preprocessing steps:
1. Lower case all text
2. De-emoji the text
3. Remove all hashtags, mentions, and urls
4. Remove all non-alphabet characters except the followng punctuations: period, exclamation mark, and question mark

Hints: 
- For step 2 (de-emoji), consider using the python [emoji](https://carpedm20.github.io/emoji/docs/) library. The `emoji.demojize` method will convert all emojis to plain text. The `emoji` library is installed in cell [52].
- Follow the processing steps in order. For example calling nltk's word_tokenize before removing hashtags and mentions will end up creating seperate tokens for @ and # and cause problems.

To get full credit for this task, the Test 1b must pass. Only modify the  cell containing the `preprocess_tweet` function and do not alter the testing code block. 

After you are satisfied with your code, run the tests. code to ensure your function works as expected. This cell will also create a new column called `cleaned_tweet` and apply the `preproces_tweet` function to all the examples in the dataset. 

In [6]:
import emoji, string, re
!pip install tweet-preprocessor
import preprocessor.api as p #Ref: https://pypi.org/project/tweet-preprocessor/

def preprocess_tweet(tweet: str) -> str:
  """
  Function takes a raw tweet and performs the following processing steps:
  1. Lower case all text
  2. De-emoji the text
  3. Remove all hashtags, mentions, and urls 
  4. Remove all non-alphabet characters except the followng punctuations: period, exclamation mark, and question mark
  """
  tweet = tweet.lower()
  tweet = emoji.demojize(tweet)
  tweet = p.clean(tweet)
  tweet = re.sub(r'[^\w\s\.\?\!]','',tweet).replace('_','')
  tweet = tweet.lstrip()
  return tweet 

test_tweet = "'@tateliverpool #BobandRoberta: I am angry more artists that have a profile are not speaking up! #foundationcourses 😠'"

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


### Task 1b Test
Run the cell below to evaluate your code. To get full credit for this task, your code must pass all tests. Any alteration of the testing code will automatically result in 0 points. 

In [7]:
# Do NOT modify the code below. 
# Create new column with cleaned tweets. We will use this for the subsequent tasks
data["cleaned_tweet"] = data["tweet"].apply(preprocess_tweet)

# Test 1b 
test_tweet = "'@tateliverpool #BobandRoberta: I am angry more artists that have a profile are not speaking up! #foundationcourses 😠'"
clean_tweet = "i am angry more artists that have a profile are not speaking up! angryface"
print(f"Test 1b: {preprocess_tweet(test_tweet) == clean_tweet}")

Test 1b: True


### Task 1c. Generating Evaluation Splits [2 points]
Finally, we need to split our data into a train, validation, and test set. We will split the data using a 60-20-20 split, where 60% of our data is used for training, 20% for validation, and 20% for testing. As the dataset is heaviliy imbalanced, make sure you stratify the dataset to ensure that the label distributions across the three splits are roughly equal. 

Store your splits in the variables `train`, `val`, and `test` respectively. 

Hints:
- Use the [`train_test_split`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) function for this task. You'll have to call it twice to get the validation split. 
- Set the random state so the sampling can be reproduced (we use 2023 for our random state)
- Use the `stratify` parameter to ensure representative label distributions across the splits. 

In [8]:
from sklearn.model_selection import train_test_split

# Your code here
train, test = train_test_split(data, train_size = 0.6, random_state = 2023, stratify = data['label'], shuffle = True)
val, test = train_test_split(test, test_size = 0.5, random_state = 2023, stratify = test['label'], shuffle = True)

## Task 2: Naive Baseline Using a Rule-based Classifier [10 points]

Now that we have a dataset, let's work on developing some solutions for emotion classification. We'll start with implementing a simple rule-based classifier which will also serve as our naive baseline. Emotive language (e.g. awesome, feel great, super happy) can be a strong signal as to the overall emotion being by the tweet. For each emotion in our label space (happy, surprised, sad, angry) we will generate a set of words and phrases that are often associated with that emotion. At classification time, the classifier will calculate a score based on the overlap between the words in the tweet and the emotive words and phrases for each of the emotions. The emotion label with the highest overlap will be selected as the prediction and if there is no match the "no-emotion" label will be predicted. We can break the implementation of this rules-based classifier into three steps:
1. Emotive language extraction from train examples 
2. Developing a scoring algorithm
3. Building the end-to-end classification flow 

### Task 2a. Emotive Language Extraction [4 points] 
For this task you will generate a set of unigrams and bigrams that will be used to predict each of the labels. Using the training data you will need to extract all the unique unigrams and bigrams associated with each label (excluding no-emotion). Then you should ensure that the extracted terms for each emotion label do not appear in the other lists. In the real world, you would then manually curate the generated lists to ensure that associated words were useful and emotive. For the assignment, you won't be required to further curate the generated lists.

Once you've identified the appropiate terms, save them as lists stored in the following environment variables: `happy_words`, `surprised_words`, `sad_words`,and `angry_words`. To get full credit for this section, ensure all 2a Tests pass. 

Hints
- We suggest you use Python's [set methods](https://realpython.com/python-sets/) for this task.
- NLTK has a function for extracting [ngrams](https://www.nltk.org/api/nltk.util.html?highlight=ngrams#nltk.util.ngrams). This function expects a list of tokens as input and will output tuples which you'll need to reconvert into strings. 

In [9]:
from nltk.metrics.association import UNIGRAMS
from torch._C import NoneType
# Your code here
from typing import List
from nltk.util import ngrams

# 1. Extract all terms associated with each label
def extract_words(examples: List[str]) -> List[str]:
  """
  Given a list of tweets, return back the unigrams and bigrams found
  across all the tweets. 
  """
  # extracted_words = []
  # for i in examples:
  unigrams = (set(ngrams(nltk.word_tokenize(examples),1)))
  bigrams = (set(ngrams(nltk.word_tokenize(examples), 2)))
  extracted_words = unigrams.union(bigrams)

  return extracted_words

happy_words = set()
sad_words = set()
angry_words = set()
surprise_words = set()

np_train = np.array(train)
for i in np_train:
  if i[2] == 'happy':
    happy_words.update(extract_words(i[3]))
  elif i[2] == 'sad':
    sad_words.update(extract_words(i[3]))
  elif i[2] == 'angry':
    angry_words.update(extract_words(i[3]))
  elif i[2] == 'surprise':
    surprise_words.update(extract_words(i[3]))
  
happy_words = happy_words.difference(angry_words, sad_words, surprise_words)
sad_words = sad_words.difference(angry_words, happy_words, surprise_words)
angry_words = angry_words.difference(happy_words, sad_words, surprise_words)
surprise_words = surprise_words.difference(angry_words, sad_words, happy_words)

### Task 2a Tests
Run the cell below to evaluate your code. To get full credit for this task, your code must pass all tests. Any alteration of the testing code will automatically result in 0 points. 

In [10]:
# Check sets are non-empty
print("Checking sets are not empty: ")
print(f"Happy words count: {len(happy_words)}, {len(happy_words) > 0}")
print(f"Sad words count: {len(sad_words)}, {len(sad_words) > 0}")
print(f"Angry words count: {len(angry_words)}, {len(angry_words) > 0}")
print(f"Surprise words count: {len(surprise_words)}, {len(surprise_words) > 0}")

# Checks sets are disjoint 
union1 = sad_words.union(angry_words, surprise_words)
union2 = happy_words.union(surprise_words, angry_words) 
union3 = surprise_words.union(happy_words, sad_words)
union4 = angry_words.union(happy_words, sad_words) 

print("\nChecking sets are all disjoint:")
print(f"Happy words disjoint: {happy_words.isdisjoint(union1)}")
print(f"Sad words disjoint: {sad_words.isdisjoint(union2)}")
print(f"Angry words disjoint: {angry_words.isdisjoint(union3)}")
print(f"Surprise words disjoint: {surprise_words.isdisjoint(union4)}")

Checking sets are not empty: 
Happy words count: 7100, True
Sad words count: 415, True
Angry words count: 762, True
Surprise words count: 354, True

Checking sets are all disjoint:
Happy words disjoint: True
Sad words disjoint: True
Angry words disjoint: True
Surprise words disjoint: True


### Task 2b. Scoring using set overlaps [2 points]

Next we will implement to scoring algorithm. Our score will simply be the count of overlapping terms between tweet text and emotive terms. For this task, finish implementing the code below. To get full credit, ensure Test 2b. is successful. 

In [11]:
sample_words = {'cat', 'hat', 'mat', 'bowling', 'bat'}
sample_tweet1 = "that cat is super cool sitting on the mat" 
sample_tweet2 = "the man in the bowling hat sat on the cat"
sample_tweet3 = "the quick brown fox jumped over the lazy dog"

def score_tweet(tweet, set_words):
  tweet = tweet.split()
  count = 0
  new_set = set()
  if isinstance(set_words, list):
    for i in set_words:
      for v in i:
        new_set.add(v)
    set_words = new_set
  for i in tweet:
    if i in set_words:
      count += 1
  return count


print(f"Test 1: {score_tweet(sample_tweet1, sample_words) == 2}")
print(f"Test 2: {score_tweet(sample_tweet2, sample_words) == 3}")
print(f"Test 3: {score_tweet(sample_tweet3, sample_words) == 0}")

Test 1: True
Test 2: True
Test 3: True


### 2c. Rule-based classification [4 points] 
Let put together our rules-based classfication system. Fill out the logic in the `simple_clf`. Given a tweet, `simple_clf` will generate the overlap score
for each of emotion labels and return the emotion label with the highest score. If there is no match amongst the emotions, the classifier will return 'no-emotion'.

To get full credit for this section, your average F1 score most be greater than 0.

In [12]:
def simple_clf(tweet: str) -> str:
  """
  Given a tweet, calculate all the emotion overlap scores.
  Return the emotion label which has the largest score. If
  overlap score is 0, return no-emotion. 
  """
  # Your code here 
  happy_score = 0
  sad_score = 0
  angry_score = 0
  surprise_score = 0
  happy_score = score_tweet(tweet, [item for item in happy_words])
  sad_score += score_tweet(tweet, [item for item in sad_words])
  angry_score += score_tweet(tweet, [item for item in angry_words])
  surprise_score += score_tweet(tweet, [item for item in surprise_words])

  if happy_score > max(sad_score, angry_score, surprise_score):
    return "happy"
  elif sad_score > max(happy_score, angry_score, surprise_score):
    return "sad"
  elif angry_score > max(happy_score, sad_score, surprise_score):
    return "angry"
  elif surprise_score > max(happy_score, sad_score, angry_score):
    return "surprise"
  else:
    return "no-emotion"


After finishing the above section, let's evaluate our how model did.

In [13]:
from sklearn.metrics import classification_report

preds = test["cleaned_tweet"].apply(simple_clf)
print(classification_report(test["label"], preds))

              precision    recall  f1-score   support

       angry       0.21      0.21      0.21        14
       happy       0.42      0.91      0.58       228
  no-emotion       0.79      0.23      0.35       357
         sad       0.00      0.00      0.00         7
    surprise       0.00      0.00      0.00         7

    accuracy                           0.48       613
   macro avg       0.28      0.27      0.23       613
weighted avg       0.62      0.48      0.42       613



## Task 3. Machine learning w/ grammar augmented features [10 points]

Now that we have a naive baseline, let's build a more sophisticated solution using machine learning. Up to this point, we have only considered the words in the tweet as our primary features. The rules-based approach is a very simple bag-of-words classifier. Can we improve performance if we provide some additional linguistic knowledge?

For Task 3 you will do the following:
- Generate part-of-speech features our tweets
- Train two different machine learning classifiers, one with linguistic features and one without
- Evaluate the trained models on the test set

### Task 3a. Grammar Augmented Feature Generation [3 points]
For this task, we will be generating part-of-speech tags for each token in our tweet. Additionally we'll lemmatize the text as well. We will directly include the POS information by appending the tag to the lemma of word itself. For example:
```
Raw Tweet: I am very angry with the increased prices.
POS Augmented Tweet: I-PRP be-VBP very-RB angry-JJ with-IN the-DT increase-VBN price-NNS .-.
```

Complete the `generate_pos_features` using the Spacy library. Once you have an implementation that works, we'll update the `train` and `test` dataframes with a new column called `tweet_with_pos` which contains the output of the `generate_pos_features` method.

In [14]:
import spacy 
from tqdm.notebook import tqdm
nlp = spacy.load("en_core_web_sm")

def generate_pos_features(tweet: str) -> str:
  """
  Given a tweet, return the lemmatized tweet augmented
  with POS tags.
  E.g.:
  Input: "cats are super cool."
  output: "cat-NNS be-VBP super-RB cool-JJ .-."
  """
  doc = nlp(tweet)
  new_tweet = ""
  for token in doc:
    new_tweet += f"{token.lemma_}-{token.tag_} "

  new_tweet = new_tweet.strip()
  return new_tweet

sample_tweet = "I hate action movies"
generate_pos_features(sample_tweet)

'I-PRP hate-VBP action-NN movie-NNS'

In [15]:
# Once you have the code working above run this cell.
train["tweet_with_pos"] = train["cleaned_tweet"].apply(generate_pos_features)
test["tweet_with_pos"] = test["cleaned_tweet"].apply(generate_pos_features)

### Task 3a Tests
Run the cell below to evaluate your code. To get full credit for this task, your code must pass all tests. Any alteration of the testing code will automatically result in 0 points. 

In [16]:
sample_texts = [
    ("i am super angry", "I-PRP be-VBP super-RB angry-JJ"),
    ("That movie was great", "that-DT movie-NN be-VBD great-JJ"),
    ("I hate action movies", "I-PRP hate-VBP action-NN movie-NNS")
]
for i, text in enumerate(sample_texts):
  print(f"Test {i+1}: {generate_pos_features(text[0]) == text[1]}")

Test 1: True
Test 2: True
Test 3: True


### Task 3b. Model Training [5 points]
Next we will train two seperate RandomForest Classifier models. For this task you will generate two sets of input features using the `TfidfVectorizer`. We generate Tfidf statistic on the`cleaned_tweet` and the `tweet_with_pos` columns. 

Once you've generated your features, train two different Random Forest classifiers with the generated features and generate the predictions on the test set for each classifier.

In [17]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.ensemble import RandomForestClassifier

# Your code here
tfidf_ct = TfidfVectorizer(
    stop_words="english",  
    max_df=.6)
tfidf_pos = TfidfVectorizer(
    stop_words="english",
    max_df=.6)
ct_train = tfidf_ct.fit_transform(train["cleaned_tweet"])
pos_train = tfidf_pos.fit_transform(train["tweet_with_pos"])

ct_test = tfidf_ct.transform(test["cleaned_tweet"])
pos_test = tfidf_pos.transform(test["tweet_with_pos"])

CT_Forest = RandomForestClassifier()
CT_Forest.fit(ct_train.toarray(), train["label"])

POS_Forest = RandomForestClassifier()
POS_Forest.fit(pos_train.toarray(), train["label"])

ct_preds = CT_Forest.predict(ct_test)

pos_preds = POS_Forest.predict(pos_test)

### Task 3c. [2 points]
Generate classification reports for both models. Print the reports below. In a few sentences (no more than 100 words) explain which features were the most effective and why you think that's the case?

In [18]:
from sklearn.metrics import classification_report

# Classification Report for Tfidf features
print("Classification report for TFIDF features")
# Your code here
print(classification_report(test["label"], ct_preds))
# Classfication Report for POS features 
print("Classification report for TFIDF w/ POS features")
# Your code here
print(classification_report(test["label"], pos_preds))

Classification report for TFIDF features
              precision    recall  f1-score   support

       angry       0.50      0.14      0.22        14
       happy       0.88      0.69      0.77       228
  no-emotion       0.78      0.94      0.86       357
         sad       0.00      0.00      0.00         7
    surprise       0.00      0.00      0.00         7

    accuracy                           0.81       613
   macro avg       0.43      0.36      0.37       613
weighted avg       0.79      0.81      0.79       613

Classification report for TFIDF w/ POS features
              precision    recall  f1-score   support

       angry       0.33      0.07      0.12        14
       happy       0.74      0.70      0.72       228
  no-emotion       0.78      0.87      0.82       357
         sad       0.00      0.00      0.00         7
    surprise       0.00      0.00      0.00         7

    accuracy                           0.77       613
   macro avg       0.37      0.33      0.3

/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classificat

## Task 4. Transfer Learning with DistilBERT [10 points]

For this task you will finetune a pretrained language model (DistilBERT) using the huggingface `transformers` library. For this task you will need to:
- Encode the tweets using the BERT tokenizer
- Create pytorch datasets for for the train, val and test datasets
- Finetune the distilbert model for 5 epochs
- Extract predictions from the model's output logits and convert them into the emotion labels.
- Generate a classification report on the predictions.

Ensure you are running the notebook in Google Colab with the gpu runtime enabled for this section.

In [19]:
from sklearn.preprocessing import LabelEncoder
import torch
torch.cuda.is_available()
from torch.utils.data import Dataset
from transformers import AutoTokenizer
from transformers import AutoModelForSequenceClassification
from transformers import Trainer
from transformers import TrainingArguments


# Your Code here
tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")

#Reference = CT5121 Lab Session Week 01
class SentimentDataset(Dataset):
    
    def __init__(self, encodings: dict):  
        self.encodings = encodings
        
    def __len__(self) -> int:
        return len(self.encodings["input_ids"])
    
    def __getitem__(self, idx: int) -> dict:
        e = {k: v[idx] for k,v in self.encodings.items()}
        return e 

train["label"] = train["label"].replace({"surprise":0, "happy":1, "no-emotion":2, "sad":3, "angry": 4})
val["label"] = val["label"].replace({"surprise":0, "happy":1, "no-emotion":2, "sad":3, "angry": 4})
test["label"] = test["label"].replace({"surprise":0, "happy":1, "no-emotion":2, "sad":3, "angry": 4})

train_encodings = tokenizer(
    train["cleaned_tweet"].tolist(), 
    padding=True,
    max_length=128,
    return_tensors="pt",
    truncation=True
)
train_encodings["labels"] = torch.tensor(train["label"].tolist())
train_dataset = SentimentDataset(train_encodings)


val_encodings = tokenizer(
    val["cleaned_tweet"].tolist(), 
    padding=True,         
    max_length=128,
    return_tensors="pt",
    truncation=True
)
val_encodings["labels"] = torch.tensor(val["label"].tolist())
val_dataset = SentimentDataset(val_encodings)



test_encodings = tokenizer(
    test["cleaned_tweet"].tolist(), 
    padding=True,
    max_length=128,
    return_tensors="pt",
    truncation=True
)
test_y = test["label"].tolist()
test_dataset = SentimentDataset(test_encodings)

bert = AutoModelForSequenceClassification.from_pretrained("distilbert-base-uncased", num_labels = 5)

#Reference = CT5121 Lab Session Week 01
training_args = TrainingArguments(
    output_dir="./results",
    num_train_epochs=5,
    evaluation_strategy="epoch",
    save_strategy="epoch",
    lr_scheduler_type='cosine',
    per_device_train_batch_size = 32,
    per_device_eval_batch_size = 32, 
    fp16=True,
)

trainer = Trainer(
    bert,
    training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
)

trainer.train()

preds = trainer.predict(test_dataset)

Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertForSequenceClassification: ['vocab_layer_norm.bias', 'vocab_projector.weight', 'vocab_transform.bias', 'vocab_layer_norm.weight', 'vocab_projector.bias', 'vocab_transform.weight']
- This IS expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier

***** Running Evaluation *****
  Num examples = 612
  Batch size = 32
Saving model checkpoint to ./results/checkpoint-58
Configuration saved in ./results/checkpoint-58/config.json
Model weights saved in ./results/checkpoint-58/pytorch_model.bin
***** Running Evaluation *****
  Num examples = 612
  Batch size = 32
Saving model checkpoint to ./results/checkpoint-116
Configuration saved in ./results/checkpoint-116/config.json
Model weights saved in ./results/checkpoint-116/pytorch_model.bin
***** Running Evaluation *****
  Num examples = 612
  Batch size = 32
Saving model checkpoint to ./results/checkpoint-174
Configuration saved in ./results/checkpoint-174/config.json
Model weights saved in ./results/checkpoint-174/pytorch_model.bin
***** Running Evaluation *****
  Num examples = 612
  Batch size = 32
Saving model checkpoint to ./results/checkpoint-232
Configuration saved in ./results/checkpoint-232/config.json
Model weights saved in ./results/checkpoint-232/pytorch_model.bin
***** Runni

In [20]:
predictions = []
for i, v in enumerate(preds[0]):
  predictions.append(np.argmax(v))
print(classification_report(test["label"], predictions))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00         7
           1       0.87      0.87      0.87       228
           2       0.87      0.91      0.89       357
           3       1.00      0.14      0.25         7
           4       0.60      0.43      0.50        14

    accuracy                           0.87       613
   macro avg       0.67      0.47      0.50       613
weighted avg       0.86      0.87      0.86       613



## Task 5. Model Recommendation [5 points]
In a paragraph (no more than 250 words) answer the following questions:
1. Which of the implemented models would you recommend and why? 
2. Compare the metrics for each models implemted (Rules-Based, Machine Learning w/ POS features, and DistilBERT). What are the pros and con for each model (consider performance both macro performance and label specifc metrics and the computational requirements). 

## Answers
1. After seeing all the three models fail on the labels "sad" and "surprise" due to its lack of training data, I'd say it would go for BERT Transformer route as it still gives us better performance score compared to other two, due to the reason that it provides good weights and bias to the internal nodes on which it determines the result, as compared to RandomForestClassifier and the Rule-based model which has fixed set of defined functions and no learning or shifting of the parameters is taking place. It takes the least time out of the three but performs with very underwhelming results out of the three.
<br/>

2. 
* Rule-Based model implemented has a fixed set of words which shows us a bag of words implementation of an NLP task, which can be good if we have a small program allocated for a smaller scale use, but lacks where a word can mean different things in correlation to others or the sentence itself. The model gives us an accuracy of 48% which is very average considering the training and testing size.

* Machine Learning w/ POS features using RandomForestClassifier is still an ideal choice with accuracy around 77% and w/o POS accuracy of 81% which still gives a good ideology of the fact that we can still use them to opt, and can change the hyperparameters or can even do better preprocessing work or use a good lexical database to support our database for preprocessing, to crux the results to even better scale, while it takes considerably less time than the Transformers and comes closer to it.

* Transformers use GPU to process the data, train and run them with a considerably high accuracy out of the tree at a whopping 87% which is a good performance score in a broad subject like NLP, but requires a good knowledge of the subject and even a bigger corpus of data to transform and encode the data. It becomes good as the data is transformed into numbers because a system works well if numbers are provided to it for processing, and that's what is achieved here. The weights and biases for each node in the layers are well set according to the inputs provided and it is a big reason for this method to work the best, although a lot is required to set metrics using PyTorch.